packages/client/src/emitter.ts

```diff
@@ -0,0 +1,56 @@
+import { uuid } from '@botpress/messaging-base'
+
+export class Emitter<T extends { [key: string]: any }> {
+  private listeners: { [eventId: string]: (((clientId: uuid, arg: any) => Promise<any>) | ((arg: any) => void))[] } = {}
+
+  public on<K extends keyof T>(
+    event: K,
+    listener: ((clientId: uuid, arg: T[K]) => Promise<any>) | ((arg: T[K]) => void)
+  ) {
+    const listeners = this.listeners[event as string]
+    if (!listeners) {
+      this.listeners[event as string] = [listener]
+    } else {
+      listeners.push(listener)
+    }
+  }
+
+  protected async emit<K extends keyof T>(event: K, clientId: uuid, arg: T[K]): Promise<boolean> {
+    const listeners = this.listeners[event as string]
+    if (listeners?.length) {
+      for (const listener of listeners) {
+        await listener(clientId, arg)
+      }
+      return true
+    } else {
+      return false
+    }
+  }
+}
+
+export class ProtectedEmitter<T extends { [key: string]: any }> {
```

Comment:
Not sure I understand what this is for? Why create two implementations of an Emitter class?

Reply:
One has two arguments it sends out with events. The other only has one.

Reply:
This is to handle clientId as the first argument of callbacks